README.md
# SystaComfortPrometheusExporter
Intercepts UDP Status Packets from a SystaComfort Heating controller and makes them available to Prometheus

Die Deutsche Übersetzung findet ihr [hier](https://github.com/xgcssch/SystaComfortPrometheusExporter/blob/main/README_de.md).

# Motivation
If you are an owner of a *SystaComfort* Heating from [Paradigma](https://www.paradigma.de/), then you have limited possibilities to monitor the
device: You can use a mobile app called [S-Touch](https://www.paradigma.de/produkte/regelungen/s-touch-app/) which just exposes the current values or minimalistic graphics or an webbased remote portal. There are neither Open APIs nor interfaces to common monitoring or control tools like MQTT.

As a result, Klaus Schmidinger decoded the protocol which every *SystaComfort* Controler sends to the Remote portal. He developed a perl script which decoded the data and produced RRD graphs.

As time goes by and technology advances further, i decided to rewrite the perl script and build an exporter for [Prometheus](https://prometheus.io/), a time-series Database specialized in storing metrics. The data contained in Prometheus can later be viewed and analyzed in tools like [Grafana](https://grafana.com).
# *SystaComfort* and Modbus TCP
This project uses a complicated and undocumented method to get access to the monitored data. A much better approach would be to use an offically supported interface. Fortunately this way is open to users of the *SystaComfort* Controller with a hardware version starting at Revision 2 - manufacturing date ~2016. The Revision can easyly be determined with a look at the board. Revision 1 has just one *LAN* connector. Starting with Revision 2 it has two connectors. If you are lucky to have this newer version, you can query and control the unit with an *Modbus TCP* Controller. 

Further information for the *SystaComfort* *Modbus TCP* interface can be found [here](https://mam-prod.paradigma.de/pinaccess/pinaccess.do?pinCode=6om1uCSh0OKd). To use the interface, a software upgrade to the latest version may be required.

Thanks to Boris Bartenstein of *Paradigma* for sharing this information.

# Requirements
Before you start, be sure you need some things:
1. A *SystaComfort* Controller configured to send data unencrypted to the Paradigma Remote Portal.
1. A configurable Router where you can change the IP Address of an official DNS entry to a local IP in your network
1. Some kind of server, where you can run this program, which converts the *SystaCoomfort* data into the format used by Prometheus. This program has to run continuously to provide the data. 
1. A permanently running Prometheus instance.
1. A Grafana instance to get the data visualized

All of these software components can be run on small devices like a [Rasperry Pi](https://www.raspberrypi.org/) or a NAS Device like Synology DiskStation capable of running Docker Container.

## Versions known to work
The following versions of the *SystaComfort* Controller are known to have worked with this hardware/software combinations:
- HW 1 - V1.24
- HW 1 - V1.26
- HW 2 - V2.14
# Installation
## Configure your DNS
The DNS Name `paradigma.remoteportal.de` must point to the device where the exporter is running. In this sample this is the Address `192.168.100.4`.

Every internet router has it's own method to set such an entry. Some models are even incapble of overriding DNS addresses. If you own such device, there is no easy way to overcome the problem. One way is to install a local DNS server in your network, override the DNS Name there and only point the SystaComfort to that specific DNS Server. Unfortunatly this is not an easy thing to do and falls outside the scope of this document.
See here for doing this with an [Rasperry Pi](https://www.deviceplus.com/raspberry-pi/how-to-use-a-raspberry-pi-as-a-dns-server/) or with a [Container](https://github.com/opsxcq/docker-dnsmasq)

But if you were able to spoof the address, you should see an output like this if you query the address:
```
luna $ dig paradigma.remoteportal.de

; <<>> DiG 9.11.3-1ubuntu1.13-Ubuntu <<>> paradigma.remoteportal.de
;; global options: +cmd
;; Got answer:
;; ->>HEADER<<- opcode: QUERY, status: NOERROR, id: 55342
;; flags: qr rd ad; QUERY: 1, ANSWER: 1, AUTHORITY: 0, ADDITIONAL: 0
;; WARNING: recursion requested but not available

;; QUESTION SECTION:
;paradigma.remoteportal.de.     IN      A

;; ANSWER SECTION:
paradigma.remoteportal.de. 0    IN      A       192.168.100.4

;; Query time: 3 msec
;; SERVER: 172.24.176.1#53(172.24.176.1)
;; WHEN: Sun Oct 25 00:34:00 CEST 2020
;; MSG SIZE  rcvd: 84
```
If you are on a windows machine, use `nslookup paradigma.remoteportal.de` to check.
## Configure *SystaComfort* Controller
There is a *Serviceprogramm Paradigma* which is normally used by maintainance personal. Below the menu group entry *Options* there is an entry called *Networksettings SystaComfort II*. When selected, it displays the Network settings of the controller. In the lower half there is a property sheet with two pages: the second is labeled *Portal*. Select this page. There are two checkboxes: the first, *Remoteportal aktiv* must be checked to enable the data transmissions. The second, *Verschlüsselung aktiv* should not be checked, as we have no information how to decrypt the data.

![Screenshot of the network configuration](https://github.com/xgcssch/SystaComfortPrometheusExporter/raw/main/doc/assets/NetworkSettings.png)

# As a docker container
The easiest way to install the exporter if you already have a device with *Docker* running, is to use a container.

Just issue
```
docker run -d --name systacomfortexporter \
    -p 22460:22460/udp \
    -p 2112:2112/tcp \
    xgcssch/systacomfortprometheusexporter
```
this will run the exporter and expose both the UDP endpoint for the heating controller and the TCP endpoint for Prometheus.

Remember that the address of this machine is the one where `paradigma.remoteportal.de` should point to!
## As a program running on a server machine
Download the binary for your platform from the *Releases+ section of this project.

Just run the executable:
```
luna $ ./SystaComfortPrometheusExporter
2020/10/25 11:23:14 Starting ...
```
Accessing the metrics endpoint with curl should show the current metrics:
```
luna $ curl http://192.168.100.4:2112/metrics
# HELP systacomfort_boiler_active_info Is the boiler is running
# TYPE systacomfort_boiler_active_info gauge
systacomfort_boiler_active_info 0
# HELP systacomfort_boiler_circulationpump_info Is the boiler circulation pump running
# TYPE systacomfort_boiler_circulationpump_info gauge
systacomfort_boiler_circulationpump_info 1
...
```

### Windows Service
To run the program as a windows service, use [NSSM](https://nssm.cc/)
# Prometheus Installation
Installation instructions can be found at the [Prometheus](https://prometheus.io/docs/prometheus/latest/installation/) website.

I would recommend using a docker container too. You will need a configuration file which points to the machine running the exporter like this:
```
global:
  evaluation_interval: 1m
  scrape_interval: 1m
  scrape_timeout: 10s
scrape_configs:
- job_name: systaComfort
  static_configs:
  - targets:
    - 192.168.100.4:2112
```
Make sure you put the data directory on persistent storage, otherwise the recorded data is lost when the container is removed!

Example:
```
docker run -d --name prometheus \
    -p 9090:9090 \
    -v /var/data/prometheus/data:/prometheus \
    -v /var/data/prometheus/config/prometheus.yml:/etc/prometheus/prometheus.yml \
    prom/prometheus
```
If everything works, you can find the Prometheus UI at `http://192.168.100.4:9090`. Below the menu entry *Status* -> *Targets* there should be a list  of scraped hosts. The host running the SystaComfortExporter should show up as 'State=UP*.
# Grafana Installation
Installation instructions can be found at the [Grafana](https://grafana.com/docs/grafana/latest/) website.
Again i would recommend using a docker container. 

As with the prometheus container, you will need a volume mount to make your configuration persistent.
```
docker run -d --name grafana \
    -p 3000:3000 \
    -v /var/data/grafana:/var/lib/grafana \
    grafana/grafana
```
To further configure *Grafana* follow these steps:
1. Login into Grafana at `http://192.168.100.4:3000`. Initial username and password are `admin`.
1. Goto `Configuration` -> `Datasources`, and press the `Add Datasource` button. Select `Prometheus` and enter `http://192.168.100.4:9090` as the URL. Press `Test and Save`.
1. Download the [dashboard template](https://github.com/xgcssch/SystaComfortPrometheusExporter/blob/main/assets/grafana/SystaComfort-Dashboard.json) from Github.
1. Continue to `Create`-> `Import` and select the downloaded file for import. Choose `Prometheus` as the datasource.
1. *Grafana* should display the data.

![Screenshot of the Grafana dashboard](https://github.com/xgcssch/SystaComfortPrometheusExporter/raw/main/doc/assets/GrafanaDashboard.png)


CHANGELOG.md
# Changelog
## 13.11.21
- Neues Grafana Dashboard mit einem Störungspanel.\
Das Panel hat einen *Alert* definiert, welcher bei einer Störung auslöst.\
Tipp: Grafana lässt sich sehr gut mit *Telegram* verbinden, so dass Störungsmeldungen über Pushnachrichten verschickt werden können.
- Hinweise zu der Verfügbarkeit einer Modbus TCP Schnittstelle für Hardwareversionen >= 2 ergänzt
- Werte für systacomfort_solarpanel_return_temperature_celsius und systacomfort_solarpanel_supply_temperature_celsius getauscht


cmd/SystaComfortPrometheusExporter/main.go
//
// Main entry for the SystaComfort Prometheus exporter
//

package main

import (
    "flag"

    internal "github.com/xgcssch/SystaComfortPrometheusExporter/internal/pkg/udpserver"
    "k8s.io/klog/v2"
)

var registerGoCollector = flag.Bool("registerGoCollector", false, "Register the GO collector")
var registerProcessCollector = flag.Bool("registerProcessCollector", false, "Register the process collector")
var prometheusPort = flag.Int("port", 2112, "Port to use exposing the exporter")
var prometheusURL = flag.String("url", "/metrics", "URL where the metrics are exposed")

func main() {
    klog.InitFlags(nil)
    flag.Parse()

    klog.Info("SystaComfortPrometheusExporter v0.1 starting")

    Configuration := internal.ProgramConfiguration{
        PrometheusPort:  *prometheusPort,
        PrometheusURL: *prometheusURL,
        RegisterGoCollector: *registerGoCollector,
        RegisterProcessCollector: *registerProcessCollector,
    }

    internal.StartupServer( Configuration )
}


internal/pkg/udpserver/prometheuscounter.go
//
// Derived from a perl script developed by Klaus.Schmidinger@tvdr.de

package udpserve

import (
	"github.com/prometheus/client_golang/prometheus"
)

var (
	// Außentemperatur Fühler
	systacomfortOutsideTemperatureCelsius = prometheus.NewGauge(prometheus.GaugeOpts{
		Name: "systacomfort_outside_temperature_celsius",
		Help: "The outside temperature",
	})
	// Heizung Vorlauf HK1
	systacomfortHeatercircuitSupplyTemperatureCelsius = prometheus.NewGauge(prometheus.GaugeOpts{
		Name: "systacomfort_heatercircuit_supply_temperature_celsius",
		Help: "The boiler supply temperature",
	})
	// Heizung Ruecklauf HK1
	systacomfortHeatercircuitReturnTemperatureCelsius = prometheus.NewGauge(prometheus.GaugeOpts{
		Name: "systacomfort_heatercircuit_return_temperature_celsius",
		Help: "The boiler return temperature",
	})
	// Brauchwasser TWO
	systacomfortTapwaterTemperatureCelsius = prometheus.NewGauge(prometheus.GaugeOpts{
		Name: "systacomfort_tapwater_temperature_celsius",
		Help: "The tapwater temperature",
	})
	// Puffertemperatur oben TPO
	systacomfortBufferToplayerTemperatureCelsius = prometheus.NewGauge(prometheus.GaugeOpts{
		Name: "systacomfort_buffer_toplayer_temperature_celsius",
		Help: "The temperature in the top layer of the buffer",
	})
	// Puffertemperatur unten TPU
	systacomfortBufferBottomlayerTemperatureCelsius = prometheus.NewGauge(prometheus.GaugeOpts{
		Name: "systacomfort_buffer_bottomlayer_temperature_celsius",
		Help: "The temperature in the bottom layer of the buffer",
	})
	// Zirkulation
	systacomfortCirculationTemperatureCelsius = prometheus.NewGauge(prometheus.GaugeOpts{
		Name: "systacomfort_watercirculation_temperature_celsius",
		Help: "The temperature in the water circulation",
	})
	// Heizung Vorlauf HK2
	systacomfortHeatercircuit2SupplyTemperatureCelsius = prometheus.NewGauge(prometheus.GaugeOpts{
		Name: "systacomfort_heatercircuit2_supply_temperature_celsius",
		Help: "The boiler supply temperature - Circuit 2",
	})
	// Heizung Ruecklauf HK2
	systacomfortHeatercircuit2ReturnTemperatureCelsius = prometheus.NewGauge(prometheus.GaugeOpts{
		Name: "systacomfort_heatercircuit2_return_temperature_celsius",
		Help: "The boiler return temperature",
	})
	// Raumtemperatur
	systacomfortInsideTemperatureCelsius = prometheus.NewGauge(prometheus.GaugeOpts{
		Name: "systacomfort_inside_temperature_celsius",
		Help: "The inside temperature",
	})
	// Raumtemperatur HK2
	systacomfortInside2TemperatureCelsius = prometheus.NewGauge(prometheus.GaugeOpts{
		Name: "systacomfort_inside2_temperature_celsius",
		Help: "The inside temperature - Circuit 2",
	})
	// Kollektor
	// KesselVorlauf
	systacomfortBoilerSupplyTemperatureCelsius = prometheus.NewGauge(prometheus.GaugeOpts{
		Name: "systacomfort_boiler_supply_temperature_celsius",
		Help: "The boiler supply temperature",
	})
	// KesselRuecklauf
	systacomfortBoilerReturnTemperatureCelsius = prometheus.NewGauge(prometheus.GaugeOpts{
		Name: "systacomfort_boiler_return_temperature_celsius",
		Help: "The boiler return temperature",
	})
	// BrauchwasserSoll
	systacomfortTapwaterTargetTemperatureCelsius = prometheus.NewGauge(prometheus.GaugeOpts{
		Name: "systacomfort_tapwater_target_temperature_celsius",
		Help: "The tapwater target temperature",
	})
	// InnenSoll
	systacomfortInsideTargetTemperatureCelsius = prometheus.NewGauge(prometheus.GaugeOpts{
		Name: "systacomfort_inside_target_temperature_celsius",
		Help: "The inside target temperature",
	})
	// Vorlauf Heizung Soll HK1
	systacomfortHeatercircuitTargetSupplyTemperatureCelsius = prometheus.NewGauge(prometheus.GaugeOpts{
		Name: "systacomfort_heatercircuit_target_supply_temperature_celsius",
		Help: "The target supply temperature",
	})
	// Vorlauf Heizung Soll HK2
	systacomfortHeatercircuit2TargetSupplyTemperatureCelsius = prometheus.NewGauge(prometheus.GaugeOpts{
		Name: "systacomfort_heatercircuit2_target_supply_temperature_celsius",
		Help: "The target supply temperature Circuit 2",
	})
	// 33 - Puffertemperatur Soll
	systacomfortBufferTargetTemperatureCelsius = prometheus.NewGauge(prometheus.GaugeOpts{
		Name: "systacomfort_buffer_target_temperature_celsius",
		Help: "The target temperature for the buffer",
	})
	// 34 - KesselSoll
	systacomfortBoilerTargetTemperatureCelsius = prometheus.NewGauge(prometheus.GaugeOpts{
		Name: "systacomfort_boiler_target_temperature_celsius",
		Help: "The boiler target temperature",
	})
	// 36 - Betriebsart 0=Heizprogramm 1, 1=Heizprogramm 2, 2=Heizprogramm 3, 3=Dauernd Normal, 4=Dauernd Komfort, 5=Dauernd Absenken
	systacomfortModeInfo = prometheus.NewGauge(prometheus.GaugeOpts{
		Name: "systacomfort_mode_info",
		Help: "The current mode",
	})
	// 39=Raumtemperatur normal (soll)
	systacomfortInsideNormalTargetTemperatureCelsius = prometheus.NewGauge(prometheus.GaugeOpts{
		Name: "systacomfort_inside_normal_target_temperature_celsius",
		Help: "The normal inside target temperature",
	})
	// 40=Raumtemperatur komfort (soll)
	systacomfortInsideComfortTargetTemperatureCelsius = prometheus.NewGauge(prometheus.GaugeOpts{
		Name: "systacomfort_inside_comfort_target_temperature_celsius",
		Help: "The comfort inside target temperature",
	})
	// 41=Raumtemperatur abgesenkt (soll)
	systacomfortInsideReducedTargetTemperatureCelsius = prometheus.NewGauge(prometheus.GaugeOpts{
		Name: "systacomfort_inside_reduced_target_temperature_celsius",
		Help: "The reduced inside target temperature",
	})
	// 48 - Fusspunkt
	systacomfortHeatCurveRootPointCelsius = prometheus.NewGauge(prometheus.GaugeOpts{
		Name: "systacomfort_heatcurve_rootpoint_celsius",
		Help: "The rootpoint of the heating curve",
	})
	// 50 - Steilheit
	systacomfortHeatCurveGradientCelsius = prometheus.NewGauge(prometheus.GaugeOpts{
		Name: "systacomfort_heatcurve_gradient_celsius",
		Help: "The gradient of the heating curve",
	})
	// 52 - Max. Vorlauftemperatur
	systacomfortBoilerSupplyUpperLimitCelsius = prometheus.NewGauge(prometheus.GaugeOpts{
		Name: "systacomfort_boiler_supply_upperlimit_celsius",
		Help: "The upper limit of the supply temperature",
	})
	// 53=Heizgrenze Heizbetrieb
	// 54=Heizgrenze Absenken
	// 55=Frostschutz Aussentemperatur
	// 56=Vorhaltezeit Aufheizen
	// 57=Raumeinfluss
	// 58=Ueberhoehung Kessel
	// 59=Spreizung Heizkreis
	// 60=Minimale Drehzahl Pumpe PHK
	// 62=Mischer Laufzeit
	// 70=Speicher Sollwert
	systacomfortReservoirTargetTemperatureCelsius = prometheus.NewGauge(prometheus.GaugeOpts{
		Name: "systacomfort_reservoir_target_temperature_celsius",
		Help: "The reservoir target temperature",
	})
	// 78=Raumtemperatur normal - HK2 (soll)
	systacomfortInsideNormal2TargetTemperatureCelsius = prometheus.NewGauge(prometheus.GaugeOpts{
		Name: "systacomfort_inside_normal2_target_temperature_celsius",
		Help: "The normal inside target temperature - Circuit 2",
	})
	// 79=Raumtemperatur komfort - HK2 (soll)
	systacomfortInsideComfort2TargetTemperatureCelsius = prometheus.NewGauge(prometheus.GaugeOpts{
		Name: "systacomfort_inside_comfort2_target_temperature_celsius",
		Help: "The comfort inside target temperature - Circuit 2",
	})
	// 80=Raumtemperatur abgesenkt - HK2 (soll)
	systacomfortInsideReduced2TargetTemperatureCelsius = prometheus.NewGauge(prometheus.GaugeOpts{
		Name: "systacomfort_inside_reduced2_target_temperature_celsius",
		Help: "The reduced inside target temperature - Circuit 2",
	})
	// 149=Brauchwassertemperatur normal
	// 150=Brauchwassertemperatur komfort
	// 155=Brauchwassertemperatur Schaltdifferenz
	// 158=Nachlauf Pumpe PK/LP
	// 162=Min. Laufzeit Kessel
	// 169=Nachlaufzeit Pumpe PZ
	// 171=Zirkulation Schaltdifferenz
	// 180 - Betriebszeit Kessel
	systacomfortBoilerRuntimeSeconds = prometheus.NewGauge(prometheus.GaugeOpts{
		Name: "systacomfort_boiler_runtime_seconds",
		Help: "The total runtime of the boiler",
	})
	// 181 - Anzahl Brennerstarts
	systacomfortBoilerStartsTotal = prometheus.NewGauge(prometheus.GaugeOpts{
		Name: "systacomfort_boiler_starts_total",
		Help: "The number of boiler starts",
	})
	// 182=Solare Leistung
	systacomfortSolarOutputKWh = prometheus.NewGauge(prometheus.GaugeOpts{
		Name: "systacomfort_solar_output_kwh",
		Help: "The solar output in kWh",
	})
	// 183=Tagesgewinn Solare Leistung
	systacomfortSolarOutputDayKWh = prometheus.NewGauge(prometheus.GaugeOpts{
		Name: "systacomfort_solar_output_day_kwh",
		Help: "The daily solar output in kWh",
	})
	// 184=Gesamtgewinn Solare Leistung
	systacomfortSolarOutputTotalKWh = prometheus.NewGauge(prometheus.GaugeOpts{
		Name: "systacomfort_solar_output_total_kwh",
		Help: "The total solar output in kWh",
	})
	// 220=Aktive Relais (Bitpattern)
	//   RelaisHeizkreispumpe    = 0x0001
	//   RelaisLadepumpe         = 0x0080
	//   RelaisZirkulationspumpe = 0x0100
	//   RelaisKessel            = 0x0200
	// Brenner aktiv wenn RelaisKessel && (KesselVorlauf - KesselRuecklauf > 2);

	systacomfortBoilerHeatercircuitPumpInfo = prometheus.NewGauge(prometheus.GaugeOpts{
		Name: "systacomfort_boiler_heatercircuitpump_info",
		Help: "Is the heatercircuit pump running",
	})
	systacomfortBoilerLoadpumpInfo = prometheus.NewGauge(prometheus.GaugeOpts{
		Name: "systacomfort_boiler_loadpump_info",
		Help: "Is the boiler loading running",
	})
	systacomfortBoilerCirculationPumpInfo = prometheus.NewGauge(prometheus.GaugeOpts{
		Name: "systacomfort_boiler_circulationpump_info",
		Help: "Is the boiler circulation pump running",
	})
	systacomfortBoilerActiveInfo = prometheus.NewGauge(prometheus.GaugeOpts{
		Name: "systacomfort_boiler_active_info",
		Help: "Is the boiler is running",
	})
	systacomfortBoilerTorchInfo = prometheus.NewGauge(prometheus.GaugeOpts{
		Name: "systacomfort_boiler_torch_info",
		Help: "Is the torch of the boiler active",
	})
	// Fehlerstatus (255 = OK)
	systacomfortBoilerErrorInfo = prometheus.NewGauge(prometheus.GaugeOpts{
		Name: "systacomfort_boiler_error_info",
		Help: "Errorinformation of the boiler",
	})
	// 232=Status
	systacomfortBoilerStatusInfo = prometheus.NewGauge(prometheus.GaugeOpts{
		Name: "systacomfort_boiler_status_info",
		Help: "State of the boiler",
	})
	// 156=Maximale Kollektortemperatur
	//systacomfortSolarpanelMaximumTemperatureCelsius = prometheus.NewGauge(prometheus.GaugeOpts{
	//	Name: "systacomfort_solarpanel_maximum_temperature_celsius",
	//	Help: "Daily Maxima of the solarpanel temperature",
	//})

	// Kollektortempratur TSA
	systacomfortSolarpanelTemperatureCelsius = prometheus.NewGauge(prometheus.GaugeOpts{
		Name: "systacomfort_solarpanel_temperature_celsius",
		Help: "The temperature of the liquid in solar heating panel",
	})
	// Solar Rücklauf
	systacomfortSolarpanelReturnTemperatureCelsius = prometheus.NewGauge(prometheus.GaugeOpts{
		Name: "systacomfort_solarpanel_return_temperature_celsius",
		Help: "The temperature of the liquid streaming from the solar heating panel",
	})
	// Solar Vorlauf
	systacomfortSolarpanelSupplyTemperatureCelsius = prometheus.NewGauge(prometheus.GaugeOpts{
		Name: "systacomfort_solarpanel_supply_temperature_celsius",
		Help: "The temperature of the liquid streaming into the solar heating panel",
	})
	// Außentemperatur Kollektor TAM
	systacomfortSolarpanelOutsideTemperatureCelsius = prometheus.NewGauge(prometheus.GaugeOpts{
		Name: "systacomfort_solarpanel_outside_temperature_celsius",
		Help: "The outside temperature measured on the solar heating panel",
	})
	// Maximale Kollektortemperatur
	systacomfortSolarpanelMaximumTemperatureCelsius = prometheus.NewGauge(prometheus.GaugeOpts{
		Name: "systacomfort_solarpanel_maximum_temperature_celsius",
		Help: "Maximum temperature of the liquid in the solar panal today",
	})
)

func registerCounter(registry *prometheus.Registry) {
	// Metrics have to be registered to be exposed:
	registry.MustRegister(
        systacomfortOutsideTemperatureCelsius,
        systacomfortHeatercircuitSupplyTemperatureCelsius,
        systacomfortHeatercircuitReturnTemperatureCelsius,
        systacomfortTapwaterTemperatureCelsius,
		systacomfortBufferToplayerTemperatureCelsius,
		systacomfortBufferBottomlayerTemperatureCelsius,
        systacomfortCirculationTemperatureCelsius,
		systacomfortHeatercircuit2SupplyTemperatureCelsius,
		systacomfortHeatercircuit2ReturnTemperatureCelsius,
        systacomfortInsideTemperatureCelsius,
		systacomfortInside2TemperatureCelsius,
        systacomfortBoilerSupplyTemperatureCelsius,
        systacomfortBoilerReturnTemperatureCelsius,
        systacomfortTapwaterTargetTemperatureCelsius,
        systacomfortInsideTargetTemperatureCelsius,
		systacomfortHeatercircuitTargetSupplyTemperatureCelsius,
		systacomfortHeatercircuit2TargetSupplyTemperatureCelsius,
		systacomfortBufferTargetTemperatureCelsius,
        systacomfortBoilerTargetTemperatureCelsius,
        systacomfortModeInfo,
		systacomfortInsideNormalTargetTemperatureCelsius,
		systacomfortInsideComfortTargetTemperatureCelsius,
		systacomfortInsideReducedTargetTemperatureCelsius,			
        systacomfortHeatCurveRootPointCelsius,
        systacomfortHeatCurveGradientCelsius,
        systacomfortBoilerSupplyUpperLimitCelsius,
		systacomfortReservoirTargetTemperatureCelsius,
		systacomfortInsideNormal2TargetTemperatureCelsius,
		systacomfortInsideComfort2TargetTemperatureCelsius,
		systacomfortInsideReduced2TargetTemperatureCelsius,
        systacomfortBoilerRuntimeSeconds,
        systacomfortBoilerStartsTotal,
		systacomfortSolarOutputKWh,
		systacomfortSolarOutputDayKWh,
		systacomfortSolarOutputTotalKWh,
        systacomfortBoilerHeatercircuitPumpInfo,
        systacomfortBoilerLoadpumpInfo,
        systacomfortBoilerCirculationPumpInfo,
        systacomfortBoilerActiveInfo,
        systacomfortBoilerTorchInfo,
        systacomfortBoilerErrorInfo,
        systacomfortBoilerStatusInfo,
        systacomfortSolarpanelTemperatureCelsius,
        systacomfortSolarpanelReturnTemperatureCelsius,
        systacomfortSolarpanelSupplyTemperatureCelsius,
        systacomfortSolarpanelOutsideTemperatureCelsius,
        systacomfortSolarpanelMaximumTemperatureCelsius,
    )
}